File: hackerrank/problem solving/algorithms/2 - implementation/53 - flatland space stations/oo/flatland_space_stations_oo.cs
// Source: https://www.hackerrank.com/challenges/flatland-space-stations/problem?isFullScreen=true

using System;

public class Solution
{
    public static void Main()
    {
        List<int> inputLine = _readAnIntArray();
        int nCities = inputLine.First(), _ = inputLine.Last();

        List<int> citiesWithSpaceStation = _readAnIntArray();
        FlatlandSpaceStations obj = new FlatlandSpaceStations(nCities, citiesWithSpaceStation);
        Console.WriteLine(obj.GetMaxDistanceFromSpaceStation());
    }

        private static List<int> _readAnIntArray()
        {
            return Console.ReadLine().Split().Select(int.Parse).ToList();
        }
}

public class FlatlandSpaceStations
{
    int _nCities;
    List<int> _citiesWithSpaceStation;
    int _maxDistanceFromSpaceStation;

    public FlatlandSpaceStations(int nCities, List<int> citiesWithSpaceStation)
    {
        _nCities = nCities;
        _citiesWithSpaceStation = citiesWithSpaceStation;
        _maxDistanceFromSpaceStation = 0;

        _citiesWithSpaceStation.Sort();
        _findMaxDistanceFromSpaceStation();
    }

        private void _findMaxDistanceFromSpaceStation()
        {
            for (int cityI = 0; cityI < _nCities; cityI++)
            {
                int distance = _findMinDistanceBetweenCityAndNearestSpaceStation(cityI);
                _maxDistanceFromSpaceStation = Math.Max(_maxDistanceFromSpaceStation, distance);
            }
        }

            private int _findMinDistanceBetweenCityAndNearestSpaceStation(int cityI)
            {
                int element = _findElementOrNearest(cityI, _citiesWithSpaceStation);
                int minDistanceBetweenCityAndNearestSpaceStation = Math.Abs(cityI - element);
                return minDistanceBetweenCityAndNearestSpaceStation;
            }

                private int _findElementOrNearest(int n, List<int> array) {
                    int max = array.Count - 1;
                    int min = 0;

                    int minDistance = Int32.MaxValue;
                    int nearestElement = n;

                    while (min <= max) {
                        int mid = (min + max) / 2;

                        int distance = Math.Abs(array[mid] - n);
                        if (distance < minDistance) {
                            nearestElement = array[mid];
                            minDistance = distance;
                        }

                        if (array[mid] == n)
                            return n;
                        else if (array[mid] > n)
                            max = mid - 1;
                        else
                            min = mid + 1;
                    }

                    return nearestElement;
                }

    public int GetMaxDistanceFromSpaceStation()
    {
        return _maxDistanceFromSpaceStation;
    }
}

File: hackerrank/problem solving/algorithms/2 - implementation/53 - flatland space stations/oo/flatland_space_stations_oo.java
// Source: https://www.hackerrank.com/challenges/flatland-space-stations/problem?isFullScreen=true

import java.util.Scanner;
import java.util.List;
import java.util.Arrays;
import java.util.stream.Collectors;
import java.util.Collections;

public class Solution {
    private static Scanner _scanner;

    public static void main(String[] args) {
        _scanner = new Scanner(System.in);

        final int nCities = _scanner.nextInt(), nCitiesWithSpaceStation = _scanner.nextInt();
        final List<Integer> citiesWithSpaceStation = _readAnIntArray(nCitiesWithSpaceStation);

        final FlatlandSpaceStations obj = new FlatlandSpaceStations(nCities, citiesWithSpaceStation);
        System.out.println(obj.getMaxDistanceFromSpaceStation());
    }

        private static List<Integer> _readAnIntArray(int size) {
            return Arrays.asList(new Integer[size]).stream().map(element -> _scanner.nextInt()).collect(Collectors.toList());
        }
}

    class FlatlandSpaceStations {
        int _nCities;
        List<Integer> _citiesWithSpaceStation;
        int _maxDistanceFromSpaceStation;

        public FlatlandSpaceStations(int nCities, List<Integer> citiesWithSpaceStation) {
            _nCities = nCities;
            _citiesWithSpaceStation = citiesWithSpaceStation;
            _maxDistanceFromSpaceStation = 0;

            Collections.sort(_citiesWithSpaceStation);
            _findMaxDistanceFromSpaceStation();
        }

            private void _findMaxDistanceFromSpaceStation() {
                for (int cityI = 0; cityI < _nCities; cityI++) {
                    final int distance = _findMinDistanceBetweenCityAndNearestSpaceStation(cityI);
                    _maxDistanceFromSpaceStation = Math.max(_maxDistanceFromSpaceStation, distance);
                }
            }

                private int _findMinDistanceBetweenCityAndNearestSpaceStation(int cityI) {
                    final int element = _findElementOrNearest(cityI, _citiesWithSpaceStation);
                    final int minDistanceBetweenCityAndNearestSpaceStation = Math.abs(cityI - element);
                    return minDistanceBetweenCityAndNearestSpaceStation;
                }

                    private int _findElementOrNearest(int n, List<Integer> array) {
                        int max = array.size() - 1;
                        int min = 0;

                        int minDistance = Integer.MAX_VALUE;
                        int nearestElement = n;

                        while (min <= max) {
                            final int mid = (int) Math.floor((min + max) / 2);

                            final int distance = Math.abs(array.get(mid) - n);
                            if (distance < minDistance) {
                                nearestElement = array.get(mid);
                                minDistance = distance;
                            }

                            if (array.get(mid) == n)
                                return n;
                            else if (array.get(mid) > n)
                                max = mid - 1;
                            else
                                min = mid + 1;
                        }

                        return nearestElement;
                    }

        public int getMaxDistanceFromSpaceStation() {
            return _maxDistanceFromSpaceStation;
        }
    }
